### PR TITLE
feat: Sphinx doc improvements

### PR DESCRIPTION
**Pull request by pweiber**
- Adjusted index to look better and separated each section to their respective file
- Added External Processing Client example
- Adjusted some typos
- Added references to 'source' files and internal links to the documentation

### CODE REVIEW

**Comment by pweiber** [comment]
Is it possible to check the new version here:
https://pweiber.github.io/service-extensions-samples/
If the content does not update, try refreshing your cache and it should be good, let me know your thoughts

File: docs/python/extproc/service/callout_server.rst
Comment (jstraceski) [inline docs/python/extproc/service/callout_server.rst:1]: Is it possible to split up the `address`...`server_thread_count` parameters into a separate section from the functions?

File: docs/python/readme/developing_callouts.rst [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                ...
+
+    There are a few callback methods in ``CalloutServer`` provided for developers to override:
+
+    * ``on_request_headers``: Process request headers.
+    * ``on_response_headers``: Process response headers.
+    * ``on_request_body``: Process request body.
+    * ``on_response_body``: Process response body.
+
+    These functions correspond to the ``oneof`` required field in a `ProcessingRequest <https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/ext_proc/v3/external_processor.proto#service-ext-proc-v3-processingrequest>_` and required response field of a `ProcessingResponse <https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/ext_procv3_processingresponse>_`.
```
Comment (jstraceski) [inline docs/python/readme/developing_callouts.rst:48]: These links appear to be broken on my end.

File: docs/python/readme/developing_callouts.rst [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        from grpc import ServicerContext
+        from envoy.service.ext_proc.v3.external_processor_pb2 import HeadersResponse, HttpHeaders
+
+    See `Using the proto files`_ for more details.
+
+    Each of the callback methods provides the given data type as an input parameter and expects the corresponding response to be returned. For example ``on_response_headers``:
+
+    * ``headers``: ``response_headers`` data from `ProcessingRequest <https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/ext_proc/v3/external_processor.proto#service-ext-proc-v3-processingrequest>`_.
+    * ``context``: associated grpc data.
+    * ``return``: `response_headers` data from `ProcessingResponse <https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/ext_proc/v3/external_processor.proto#service-ext-proc-v3-processingresponse>`_.
```
Comment (jstraceski) [inline docs/python/readme/developing_callouts.rst:65]: Double tildes around response_headers.

File: docs/python/readme/quick_start.rst [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        python -m venv env
+        source env/bin/activate
+
+2. **Install Packages**:
+
+    .. code-block:: shell
+
+        pip install -r requirements.txt
+
+3. **Install `buf`**:
```
Comment (jstraceski) [inline docs/python/readme/quick_start.rst:27]: I don't think that ` work in titles, these can be removed.

File: docs/python/readme/quick_start.rst [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    .. code-block:: shell
+
+        buf -v generate \
+          https://github.com/envoyproxy/envoy.git#subdir=api \
+          --path envoy/service/ext_proc/v3/external_processor.proto \
+          --include-imports
+
+    The default template file ``buf.gen.yaml`` will not generate ``pyright`` compatible proto stubs.
+    If you plan to develop callouts with a similar type checker and not just build them,
```
Comment (jstraceski) [inline docs/python/readme/quick_start.rst:43]: Can we add a line break here.

File: docs/python/readme/developing_callouts.rst [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        from extproc.service.callout_server import CalloutServer
+
+    Just from importing the server class we can make the server run by creating a new instance and calling the blocking function ``run``:
+
+    .. code-block:: python
+
+        if __name__ == '__main__':
+            CalloutServer().run()
+
+2. **Extend ``CalloutServer``**:
```
Comment (jstraceski) [inline docs/python/readme/developing_callouts.rst:29]: Here the tildes can also be removed.

**Review by jstraceski — COMMENTED** [review]
First pass :) looking good.

File: docs/python/extproc/service/callout_server.rst
Comment (pweiber) [reply docs/python/extproc/service/callout_server.rst:1]: I tried some alternatives, let me know if the current one fits what you mentioned

File: docs/python/readme/developing_callouts.rst (hunk 1 above)
Comment (pweiber) [reply docs/python/readme/developing_callouts.rst:48]: Both are working fine here on my end, could you check again with the new build?

Comment (jstraceski) [reply docs/python/readme/developing_callouts.rst:48]: I think that the `` _` `` needs to be `` `_ `` Like below.

File: docs/python/extproc/service/callout_server.rst
Comment (jstraceski) [reply docs/python/extproc/service/callout_server.rst:1]: The current format is looking good, I think that what you have now is great.

File: docs/python/readme/requirements.rst [hunk 6]
```diff
@@ -0,0 +1,11 @@
+.. _requirements:
+
+Requirements
+============
+
+* Python 3.11+
+* `buf <https://buf.build/docs/introduction>`_
+* requirements.txt
+
+.. note::
```
Comment (jstraceski) [inline docs/python/readme/requirements.rst:10]: Quick question, If I modify the README to have this in the quick_start section, I will also have to edit this file correct?

**Review by jstraceski — COMMENTED** [review]
Awesome thanks for making those changes, basically just a small nitpick left.

File: docs/python/readme/developing_callouts.rst (hunk 1 above)
Comment (pweiber) [reply docs/python/readme/developing_callouts.rst]: Fixed, now it should be good, can you check that again to see if it is working now?

File: docs/python/readme/requirements.rst (hunk 6 above)
Comment (pweiber) [reply docs/python/readme/requirements.rst:10]: Yes, that's correct, you need to edit the corresponding .rst file, I believe there is no way to pull the content from the README automatically but I'll do some research and if find something I'll let you know

File: docs/python/readme/alternative_install_methods.rst [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+----------------
+
+Install packages through the package manager:
+
+.. code-block:: shell
+
+    sudo apt-get install python3-grpcio -y
+    python -m pip install ./protodef
+
+``[!WARNING]``
```
Comment (jstraceski) [inline docs/python/readme/alternative_install_methods.rst:16]: Should this be `.. warning::`?

File: docs/python/readme/docker.rst [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+Run the image:
+
+.. code-block:: shell
+
+    docker run -P -it --network host service-callout-example-python:latest
+
+In this example, using the ``-P`` flag tells docker to connect the exposed ports to the local machine's ports.
+Setting ``--network host`` tells docker to connect the image to the ``0.0.0.0`` or ``localhost`` ip address.
+
+``[!NOTE]``
```
Comment (jstraceski) [inline docs/python/readme/docker.rst:37]: `.. note::`

**Review by jstraceski — APPROVED** [review]
Just two last nit changes. Approved.

**Review by rshriram — APPROVED** [review]

**Pull request merged**